### PR TITLE
Introduce concept of enforced atomic operations (fix #193151)

### LINKED ISSUE #193151: Backup file system provider can not do buffered write
Does this issue occur when all extensions are disabled?: Yes/No

- VS Code Version: 1.83-Insiders
- OS Version: macOS

I was verifying https://github.com/microsoft/vscode/issues/192970 and found that large file piece tree buffers are still merged before write. Turns out that backup is always doing `doWriteUnbuffered` since it's using `FileUserDataProvider`, which doesn't have `FileOpenReadWriteClose` capability so it always runs https://github.com/microsoft/vscode/blob/3b60c12380f5a9782439be309d8d2642835b8822/src/vs/platform/files/common/fileService.ts#L394-L395

and unfortunately unbuffered write always concatenates all `buffers` first 

https://github.com/microsoft/vscode/blob/3b60c12380f5a9782439be309d8d2642835b8822/src/vs/platform/files/common/fileService.ts#L1294

image

### PR DESCRIPTION
**Pull request by rebornix**
The only two file system providers being passed to `FileUserDataProvider` is either the disk one or the in memory one. The former already supports `IFileSystemProviderWithOpenReadWriteCloseCapability` thus in this PR we added the capability to the in memory file system provider

### CODE REVIEW

File: src/vs/platform/files/common/inMemoryFilesystemProvider.ts [hunk 1]
```diff
@@ -241,4 +281,10 @@ export class InMemoryFileSystemProvider extends Disposable implements IFileSyste
 			this._bufferedChanges.length = 0;
 		}, 5);
 	}
+
+	public override dispose(): void {
```
Comment (bpasero) [inline src/vs/platform/files/common/inMemoryFilesystemProvider.ts:285]:
```suggestion
	override dispose(): void {
```

**Review by bpasero — CHANGES_REQUESTED** [review]
LGTM except one code style issue.

I leave it up to @sandy081 to comment on the impact of this change. I think the idea of this provider was to explicitly enforce atomic writes on the level of the file system provider for all user data files (except backups). With your change, we now rely on the file service to use `writeFile` for this to happen. But `writeFile` is used only in specific conditions, for example when the file is small (`3 * 64kb` I believe) or when a file is changed not through an editor that has a `ITextSnapshot` but programmatically where the entire buffer is passed in (which I believe is most often the case with user data).

**Comment by sandy081** [comment]
> or when a file is changed not through an editor that has a ITextSnapshot but programmatically where the entire buffer is passed in (which I believe is most often the case with user data).

@bpasero Is it also true if programmatic save goes through `ITextFileService.save` ?

https://github.com/microsoft/vscode/blob/0b08c9ad9c758f24effb9d92aabba47eb58818a3/src/vs/workbench/services/textfile/common/textfiles.ts#L60

**Comment by bpasero** [comment]
@sandy081 no. But I think a better solution is to stop overriding the file system provider and do what I had suggested in https://github.com/microsoft/vscode/issues/193151#issuecomment-1722765736

**Comment by sandy081** [comment]
In this case, I would not suggest this change. It is important to have atomic write for user data like settings and extensions.

**Review by bpasero — APPROVED** [review]

**Comment by bpasero** [comment]
This is ready for review.

File: src/vs/platform/userData/common/fileUserDataProvider.ts [hunk 2]
```diff
@@ -23,44 +21,63 @@ import { IUriIdentityService } from 'vs/platform/uriIdentity/common/uriIdentity'
  */
 export class FileUserDataProvider extends Disposable implements
```
Comment (bpasero) [inline src/vs/platform/userData/common/fileUserDataProvider.ts:22]: The gist of changes in this file is to be a true wrapper around the passed in disk file system provider and really forward all methods verbatim, with the converted `URI`. Please check on that, it should be simpler now.

File: src/vs/platform/files/common/fileService.ts [hunk 3]
```diff
@@ -359,10 +359,18 @@ export class FileService extends Disposable implements IFileService {
 		const provider = this.throwIfFileSystemIsReadonly(await this.withWriteProvider(resource), resource);
 		const { providerExtUri } = this.getExtUri(provider);
 
+		let writeFileOptions = options;
+		if (hasFileAtomicWriteCapability(provider) && !writeFileOptions?.atomic) {
```
Comment (bpasero) [inline src/vs/platform/files/common/fileService.ts:363]: test

File: src/vs/platform/files/common/fileService.ts [hunk 4]
```diff
@@ -389,30 +397,34 @@ export class FileService extends Disposable implements IFileService {
...
 			}
 
-			// write file: unbuffered (only if data to write is a buffer, or the provider has no buffered write capability)
-			if (!hasOpenReadWriteCloseCapability(provider) || (hasReadWriteCapability(provider) && bufferOrReadableOrStreamOrBufferedStream instanceof VSBuffer)) {
-				await this.doWriteUnbuffered(provider, resource, options, bufferOrReadableOrStreamOrBufferedStream);
+			// write file: unbuffered
+			if (
+				!hasOpenReadWriteCloseCapability(provider) ||																// buffered writing is unsupported
+				(hasReadWriteCapability(provider) && bufferOrReadableOrStreamOrBufferedStream instanceof VSBuffer) ||		// data is a full buffer already
+				(hasReadWriteCapability(provider) && hasFileAtomicWriteCapability(provider) && writeFileOptions?.atomic)	// atomic write forces unbuffered write if the provider supports it
```
Comment (bpasero) [inline src/vs/platform/files/common/fileService.ts:404]: This should have been there before: an atomic write with an atomic file system provider forces an unbuffered write directly in the provider.

**Review by sandy081 — APPROVED** [review]
Good change. I liked it.

**Review by sandy081 — APPROVED** [review]
Good change. I liked it.

**Pull request merged**